node/core/pvf/src/error.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+/// A error raised during validation of the candidate.
+#[derive(Debug, Clone)]
+pub enum ValidationError {
+	/// The error was raised because the candidate is invalid.
+	InvalidCandidate(InvalidCandidate),
+	/// This error is raised due to inability to serve the request.
+	InternalError(String),
+}
+
+/// A description of an error raised during executing a PVF and can be attributed to the combination
+/// of the candidate [`polkadot_parachain::primitives::ValidationParams`] and the PVF.
+#[derive(Debug, Clone)]
+pub enum InvalidCandidate {
+	/// The failure is reported by the worker. The string contains the error message.
+	///
+	/// This also includes the errors reported by the preparation pipeline.
+	WorkerReportedError(String),
+	/// The worker has died during validation of a candidate. That may fall in one of the following
+	/// categories, which we cannot distinguish programmatically:
+	///
+	/// (a) Some sort of transient glitch caused the worker process to abort. An example would be that
+	///     the host machine ran out of free memory and the OOM killer started killing the processes,
+	///     and in order to save the parent it will "sacrifice child" first.
+	///
+	/// (b) The candidate trigerred a code path that has lead to the process death. For example,
+	///     the PVF found a way to consume unbounded amount of resources and then it either exceeded
+	///     an rlimit (if set) or, again, invited OOM killer. Another possibility is a bug in
+	///     wasmtime allowed the PVF to gain control over the execution worker.
+	///
+	/// We attribute such an event to an invalid candidate in either case.
```

Comment:
This seems mostly fair but it is worrying. For instance, it will not be a problem in backing, but if it leads the validator to issue a dispute against the candidate which is then counteracted by all other validators, the validator will be slashed.

Ambiguous child process death should be very rare, though, no?

Reply:
Yes, I suppose validators must check rlimit, system memory, etc. before running and simply not in validator mode if they lack the resource, but..

As in https://github.com/paritytech/polkadot/pull/1656 we should worry if machines behave differently here, like an adversary triggers a slash because someone set their rlimit different from everyone else.  

We should start with quite low slashing for triggering an escalation unjustly.  We'll refund such slashes of course, except there exists some risk we might've trouble reproducing, making a big mess.

Reply:
Disputes should probably have asymmetric slashing. If you dispute against something and lose, you get slashed less than if you dispute for something and lose.

Reply:
Yes exactly, initially way way less, but later maybe significant.